### PR TITLE
Add solidity-coverage package

### PR DESCRIPTION
**Pull request by aodhgan**
### Purpose for this PR
Generates a nice coverage report
as per #133

### CODE REVIEW

**Comment by aodhgan** [comment]
Will add workflow + badges shortly!

File: README.md [hunk 1]
```diff
@@ -1,6 +1,8 @@
+[![Coverage Status](https://coveralls.io/repos/github/windranger-io/windranger-solidity-template/badge.svg?branch=master)](https://coveralls.io/github/windranger-io/windranger-solidity-template?branch=master)
```
Comment (aodhgan) [inline README.md:1]: Do we need to create the windranger-io organisation on Coveralls I wonder? The docs for the Coveralls workflow plugin says the repo (and thus badge) will be created automatically from the plugin.

How do we prevent a downstream template repo from using this (windranger-solidity-template) badge link automatically?

File: .solcover.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+module.exports = {
```
Comment (aodhgan) [inline .solcover.js:1]: Doesnt look like https://www.npmjs.com/package/solidity-coverage supports a native typescript config file

**Comment by aodhgan** [comment]
Screenshot 2022-03-10 at 3 09 53 PM
Don't know why its not playing nicely with typechain! It generates the types above, but then can't find them to import them into the tests?

**Comment by CjHare** [comment]
> Screenshot 2022-03-10 at 3 09 53 PM Don't know why its not playing nicely with typechain! It generates the types above, but then can't find them to import them into the tests?

Is this a path & container issue?

**Comment by aodhgan** [comment]
Looks like https://github.com/sc-forks/solidity-coverage/issues/656#issuecomment-943634020 has the same problem. Seems like an ordering thing.

**Comment by aodhgan** [comment]
Yeah is working now with the `TS_NODE_TRANSPILE_ONLY=true` flag set in the npm run command (as described in the above linked issue)
Screenshot 2022-03-10 at 3 26 13 PM

**Comment by CjHare** [comment]
Did you checkout https://www.npmjs.com/package/@0x/sol-coverage as it has more recent activity / release?

File: .github/workflows/main-push-pull.yml [hunk 3]
```diff
@@ -131,3 +131,36 @@ jobs:
...
+        with:
+          node-version: '16'
+
+      - name: Install dependencies
+        run: npm ci
+
+      - name: Run Coverage
+        run: npm run coverage
+
+      - name: Coveralls
```
Comment (CjHare) [inline .github/workflows/main-push-pull.yml:162]: What is coveralls? what's it doing?

File: README.md [hunk 4]
```diff
@@ -50,6 +52,134 @@ npm test
 
 Setup and run instructions:
 
-- [Hardhat](./docs/tools/hardhat.md)
-- [PlantUML](./docs/tools/plantuml.md); UML diagram generation from code.
-- [Slither](./docs/tools/slither.md); Trail of Bits Solidity static analyzer.
+#### All tests
```
Comment (CjHare) [inline README.md:55]: No, keep to the new read structure ...don't replace with the old one

File: README.md (hunk 1 above)
Comment (CjHare) [reply README.md:1]: >How do we prevent a downstream template repo from using this (windranger-solidity-template) badge link automatically?

Unless authentication tokens are involved, I don't think we can prevent that 🤷

**Review by CjHare — COMMENTED** [review]
question: Should a notice be added for the lovers of private repos and coveralls code coverage restrictions?

Private repos are not free https://coveralls.io/pricing

**Comment by aodhgan** [comment]
> question: Should a notice be added for the lovers of private repos and coveralls code coverage restrictions?
Do you mean in the readme?

We could also try code climate (https://codeclimate.com/quality/) but all these seem to have revenue models which rely on subscriptions for private repos.

**Comment by bitparadigm** [comment]
@aodhgan @CjHare is this one ready? Maybe we should push last changes, approve and merge it? There were no updates on this PR for almost 2 weeks.

File: .github/workflows/main-push-pull.yml [hunk 5]
```diff
@@ -116,18 +115,45 @@ jobs:
       - name: Set up node
         uses: actions/setup-node@v3
         with:
-          node-version: '16'
+          node-version: "16"
```
Comment (CjHare) [inline .github/workflows/main-push-pull.yml:118]: question: what's the reason for switching quote types?

File: .solcover.js [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+  providerOptions: {
+    network_id: 1337,
```
Comment (CjHare) [inline .solcover.js:3]: question: are these chain ids arbitrary or must they match the hardhat config?

File: tsconfig.json [hunk 7]
```diff
@@ -20,5 +20,5 @@
   },
   "files": ["./hardhat.config.ts"],
   "include": ["./config/**/*.ts", "./scripts/**/*.ts", "./test/**/*.ts"],
-  "exclude": ["node_modules"]
+  "exclude": ["node_modules", ".solcover.js"]
```
Comment (CjHare) [inline tsconfig.json:23]: question: do you really need to explicitly exclude that top level JS?

_Aren't all files at the top level excluded by default, hence the explicit mention of the hardhat config?_

**Comment by CjHare** [comment]
@aodhgan this PR has been superseded by https://github.com/windranger-io/hardhat-contract-tools

**Pull request closed without merging**